### PR TITLE
DO-1467

### PR DESCRIPTION
**Pull request by Hedgehog18**
NOT FOR MERGE! For review and comments only.

Prototype of a model and serializer for the register of Ukrainian sanctions.

Examples of a list of sanctions against individuals: https://www.rnbo.gov.ua/files/2016/dodat_1.pdf, https://www.rnbo.gov.ua/files/2021/dodatok_1_29_2021_36.doc
Examples of a list of sanctions against legal entities: https://www.rnbo.gov.ua/files/2016/dodat_2.pdf, https://www.rnbo.gov.ua/files/2021/dodatok_2_29_2021_36.doc

All fields described in the model appear at least once in the lists of sanctions.

### CODE REVIEW

File: sanctions_register/models/ukr_sanction.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    TYPES = (
+        (UKRAINIAN_JURIDICAL_PERSON, _('Ukrainian juridical person')),
+        (FOREIGN_JURIDICAL_PERSON, _('foreign juridical person')),
+        (UKRAINIAN_PHYSICAL_PERSON, _('Ukrainian physical person')),
+        (FOREIGN_PHYSICAL_PERSON, _('foreign physical person')),
+    )
+    person_type = models.CharField(_('type of person'), choices=TYPES, max_length=3)
+    fullname = models.TextField(_('full name'), db_index=True)
+    fullname_foreign_language = models.TextField(_('full name (in the language of the country of origin)'), db_index=True)
+    date_of_birth = models.CharField(_('date of birth'), max_length=10)
```
Comment (IegorSoboliev) [inline sanctions_register/models/ukr_sanction.py:20]: maybe, DateField?

File: sanctions_register/models/ukr_sanction.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    place_of_birth = models.CharField(_('place of birth'), max_length=100)
+    place_of_registration_physical_person = models.TextField(_('place of registration physical person'))
+    place_of_residence_physical_person = models.TextField(_('place of residence physical person'))
+    professional_activity = models.TextField(_('position or professional activity of physical person'))
+    citizenship = models.CharField(_('citizenship of physical person'))
+    documents_of_person = models.TextField(_('documents of an physical person'))
+    taxpayer_code_person = models.CharField(_('taxpayer code for an physical person'), max_length=10)
```
Comment (IegorSoboliev) [inline sanctions_register/models/ukr_sanction.py:27]: maybe, IntegerField?

File: sanctions_register/models/ukr_sanction.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    place_of_location_juridical_person = models.TextField(_('place of location juridical person'))
+    office_location = models.TextField(_('location of the juridical person office'))
+    registration_date = models.CharField(_('date of registration of the juridical person'), max_length=10)
```
Comment (IegorSoboliev) [inline sanctions_register/models/ukr_sanction.py:31]: DateField?

File: sanctions_register/models/ukr_sanction.py [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    UKRAINIAN_PHYSICAL_PERSON = 'UPP'
+    FOREIGN_PHYSICAL_PERSON = 'FPP'
+    TYPES = (
+        (UKRAINIAN_JURIDICAL_PERSON, _('Ukrainian juridical person')),
+        (FOREIGN_JURIDICAL_PERSON, _('foreign juridical person')),
+        (UKRAINIAN_PHYSICAL_PERSON, _('Ukrainian physical person')),
+        (FOREIGN_PHYSICAL_PERSON, _('foreign physical person')),
+    )
+    person_type = models.CharField(_('type of person'), choices=TYPES, max_length=3)
+    fullname = models.TextField(_('full name'), db_index=True)
```
[the remark this answers is not in the document]
Comment (Hedgehog18) [reply sanctions_register/models/ukr_sanction.py:18]: Yes. I want to use the "Full name" field for both individuals and legal entities. There may be quite "long" names.

File: sanctions_register/models/ukr_sanction.py (hunk 1 above)
Comment (Hedgehog18) [reply sanctions_register/models/ukr_sanction.py:20]: +

Comment (Hedgehog18) [reply sanctions_register/models/ukr_sanction.py:20]: I agree

File: sanctions_register/models/ukr_sanction.py (hunk 2 above)
Comment (Hedgehog18) [reply sanctions_register/models/ukr_sanction.py:27]: I agree

File: sanctions_register/models/ukr_sanction.py (hunk 3 above)
Comment (Hedgehog18) [reply sanctions_register/models/ukr_sanction.py:31]: Yes ))

File: sanctions_register/models/ukr_sanction.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    place_of_location_juridical_person = models.TextField(_('place of location juridical person'))
+    office_location = models.TextField(_('location of the juridical person office'))
+    registration_date = models.CharField(_('date of registration of the juridical person'), max_length=10)
+    edprou = models.CharField(_('code EDRPOU for Ukrainian juridical person'), max_length=8, db_index=True)
+    basic_registration_number = models.TextField(_('basic registration number for foreign juridical person'))
+    okpo = models.TextField(_('all-russian classifier of enterprises and organizations'))
+    taxpayer_code_juridical = models.CharField(_('taxpayer code for an juridical person'), max_length=10)
+
+    type_of_sanctions = models.TextField(_('type of sanctions applied'))
+    date_of_start = models.CharField(_('date of entry into force of sanctions'), max_length=10)
```
[the remark this answers is not in the document]
Comment (Hedgehog18) [reply sanctions_register/models/ukr_sanction.py:38]: I disagree. The "expiration_date" field is not strictly formatted. If the term of the sanctions is, for example, a year, we can calculate the end date. But what about the sanctions that were imposed "before the return of Crimea"? Therefore, I defined the "expiration_date" field as "TextField"

File: sanctions_register/models/ukr_sanction.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (criticus) [reply sanctions_register/models/ukr_sanction.py:18]: agree, charfield would work better here

File: sanctions_register/models/ukr_sanction.py [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+        (UKRAINIAN_PHYSICAL_PERSON, _('Ukrainian physical person')),
+        (FOREIGN_PHYSICAL_PERSON, _('foreign physical person')),
+    )
+    person_type = models.CharField(_('type of person'), choices=TYPES, max_length=3)
+    fullname = models.TextField(_('full name'), db_index=True)
+    fullname_foreign_language = models.TextField(_('full name (in the language of the country of origin)'), db_index=True)
+    date_of_birth = models.CharField(_('date of birth'), max_length=10)
+    place_of_birth = models.CharField(_('place of birth'), max_length=100)
+    place_of_registration_physical_person = models.TextField(_('place of registration physical person'))
+    place_of_residence_physical_person = models.TextField(_('place of residence physical person'))
```
Comment (criticus) [inline sanctions_register/models/ukr_sanction.py:23]: charfield?

File: sanctions_register/models/ukr_sanction.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    professional_activity = models.TextField(_('position or professional activity of physical person'))
```
Comment (criticus) [inline sanctions_register/models/ukr_sanction.py:24]: charfield?

File: sanctions_register/models/ukr_sanction.py [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    citizenship = models.CharField(_('citizenship of physical person'))
```
Comment (criticus) [inline sanctions_register/models/ukr_sanction.py:25]: should this be a list of countries?

File: sanctions_register/models/ukr_sanction.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    documents_of_person = models.TextField(_('documents of an physical person'))
```
Comment (criticus) [inline sanctions_register/models/ukr_sanction.py:26]: charfield?

File: sanctions_register/models/ukr_sanction.py [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    taxpayer_code_person = models.CharField(_('taxpayer code for an physical person'), max_length=10)
+
+    place_of_location_juridical_person = models.TextField(_('place of location juridical person'))
+    office_location = models.TextField(_('location of the juridical person office'))
```
Comment (criticus) [inline sanctions_register/models/ukr_sanction.py:30]: charfield?

File: sanctions_register/models/ukr_sanction.py [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+    registration_date = models.CharField(_('date of registration of the juridical person'), max_length=10)
+    edprou = models.CharField(_('code EDRPOU for Ukrainian juridical person'), max_length=8, db_index=True)
+    basic_registration_number = models.TextField(_('basic registration number for foreign juridical person'))
```
Comment (criticus) [inline sanctions_register/models/ukr_sanction.py:33]: charfield?

File: sanctions_register/models/ukr_sanction.py [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+    okpo = models.TextField(_('all-russian classifier of enterprises and organizations'))
```
Comment (criticus) [inline sanctions_register/models/ukr_sanction.py:34]: charfield?

File: sanctions_register/models/ukr_sanction.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (criticus) [reply sanctions_register/models/ukr_sanction.py:38]: please explain. what are examples of different formats?

File: sanctions_register/models/ukr_sanction.py [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    taxpayer_code_juridical = models.CharField(_('taxpayer code for an juridical person'), max_length=10)
+
+    type_of_sanctions = models.TextField(_('type of sanctions applied'))
+    date_of_start = models.CharField(_('date of entry into force of sanctions'), max_length=10)
+    expiration_date = models.TextField(_('the expiration date of the sanctions'))
```
Comment (criticus) [inline sanctions_register/models/ukr_sanction.py:39]: date field?

File: sanctions_register/models/ukr_sanction.py (hunk 7 above)
Comment (Hedgehog18) [reply sanctions_register/models/ukr_sanction.py:24]: How to define "max_length"? 
"professional_activity" field can be either empty or have 150 or more characters.

File: sanctions_register/models/ukr_sanction.py (hunk 2 above)
Comment (Hedgehog18) [reply sanctions_register/models/ukr_sanction.py:27]: I did not find an unambiguous answer to this question, so I suggest leaving a "char" of 15 characters

File: sanctions_register/models/ukr_sanction.py (hunk 12 above)
Comment (Hedgehog18) [reply sanctions_register/models/ukr_sanction.py:34]: This is an analogue of the Ukrainian "EDRPOU", contains either 8 or 10 digits. I think there will be a better "integer".

File: sanctions_register/models/ukr_sanction.py [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+
+    place_of_location_juridical_person = models.TextField(_('place of location juridical person'))
+    office_location = models.TextField(_('location of the juridical person office'))
+    registration_date = models.CharField(_('date of registration of the juridical person'), max_length=10)
+    edprou = models.CharField(_('code EDRPOU for Ukrainian juridical person'), max_length=8, db_index=True)
+    basic_registration_number = models.TextField(_('basic registration number for foreign juridical person'))
+    okpo = models.TextField(_('all-russian classifier of enterprises and organizations'))
+    taxpayer_code_juridical = models.CharField(_('taxpayer code for an juridical person'), max_length=10)
+
+    type_of_sanctions = models.TextField(_('type of sanctions applied'))
```
[the remark this answers is not in the document]
Comment (Hedgehog18) [reply sanctions_register/models/ukr_sanction.py:37]: These data are determined by the Law of Ukraine "On Sanctions". The law contains 25 types of sanctions. The resolutions of the National Security and Defense Council do not specify the numbers of articles of the law, it is necessary to compare them with the law manually. You can make a separate table with articles of the law.

File: sanctions_register/models/ukr_sanction.py (hunk 13 above)
Comment (Hedgehog18) [reply sanctions_register/models/ukr_sanction.py:39]: No. Answered in detail in Slack.

**Pull request closed without merging**